### ISSUE TITLE
Is it possible to re-authenticate a password while a user is signed-in? 

### ISSUE THREAD

**Issue by TheVaporTrail** [aws-amplify/amplify-js#1582]
** Which Category is your question related to? **
Authentication/Cognito

** What AWS Services are you utilizing? **
Authentication, Storage, Analytics

** Provide additional details e.g. code snippets **
I would like to the user to enter their password again before I perform a dangerous action, in particular, before deleting all of the user's stored data and then the user's account (with CognitoUser.deleteUser).

Is there a way, while the user is authenticated, to submit the password to Cognito for re-verification? I don't want a password verification failure at this point to affect the authentication state.

Thank you,
David

**Comment by elorzafe** [comment]
@TheVaporTrail currently that is not possible to do. We have an open RFC with admin auth task [here](https://github.com/aws-amplify/amplify-cli/issues/766).

**Comment by mary-cheung** [comment]
This feature would be nice to have for our team too.

**Comment by Mowinski** [comment]
Yeah :) My team also need this feature :+1:

**Comment by Amaanuddin** [comment]
same here

**Comment by ozumpe** [comment]
Is there any news on this request?

**Comment by swrap** [comment]
## Hacky Way
I have not been able to find an API call to Cognito to be able to verify an authentication type. My particular case involves just a username and password. After clicking into the internals of the AWS Amplify code there is a simple way to replicate the authentication process without interfering with the currently authenticated user. Here is my particular criteria that I was trying to solve, you can adapt this solution to your own approach:
1. Verify a user without hijack the current user session
2. Use the `USER_SRP_AUTH` flow for authenticate users
```javascript
import { AuthenticationDetails, CognitoUser, CognitoUserPool } from 'amazon-cognito-identity-js'

function fakeAuth(username, password) {
	let authDetail = new AuthenticationDetails({
		Username: username,
		Password: password,
	})
	console.log('AuthDetail', authDetail)
	let cognitoUser = new CognitoUser({
		Username: authDetail.getUsername(),
		Pool: new CognitoUserPool({
			UserPoolId: <USER_POOL_INFORMATION_GOES_HERE>,
 			ClientId: <CLIENT_ID_INFORMATION_GOES_HERE>,
		}),
	})
 	console.log('Cognito User', cognitoUser)
        //For some reason when I passed in Storage in the CognitoUserPool(...) it did not
        //want to set it, so I just override it below here. When you set the storage you can
        //can actually do whatever you with it, the storage is where are the token
        //information is kept, but seeing as I don't need it, I just blackhole it
	cognitoUser.storage = {
 		getItem: (key, value) => {
			console.log('getItem', key, value)
			return ''
		},
		setItem: (key) => {
			console.log('setItem', key)
			return ''
		},
		removeItem: (key) => {
			console.log('removeItem')
		},
		clear: () => { console.log('clear') },
	}
	console.log('Cognito User', cognitoUser)
	cognitoUser.setAuthenticationFlowType('USER_SRP_AUTH')
	let promise = new Promise((resolve, reject) => {
		cognitoUser.authenticateUser(authDetail, { onSuccess: resolve, onFailure: reject })
	})
	promise.then(user => {
		console.log('Success', user)
	}).catch(e => {
		console.log('Error', e)
	})
}
```

### Downsides:
The big downside here and how it might affect your application would be that this simulates a user logging in, which means any analytic information may be mislead by the extra user logins. However, what this does offer is a practical approach to testing a users login information.

### Why not use an API call?
Like I said earlier I would have loved to use a Cognito API call for verifying a user that is already logged in, but I could not find one, so this will have to do! Let me know your thoughts.

### Further information
Most of this was taken from `Auth.js` inside the Amplify library. The reason I did not use the `Auth` class is because it is meshed in with a lot of other Amplify classes and it seemed easy just to go straight to the meet of code, which is below for reference in case you want to know how I came up with the solution:
```javascript
...
	var authDetails = new amazon_cognito_identity_js_1.AuthenticationDetails({
		Username: username,
		Password: password,
		ValidationData: validationData
	});
	if (password) {
		return this.signInWithPassword(authDetails);
	}
	else {
		return this.signInWithoutPassword(authDetails);
	}
...
AuthClass.prototype.signInWithPassword = function (authDetails) {
	var _this = this;
	var user = this.createCognitoUser(authDetails.getUsername());
	return new Promise(function (resolve, reject) {
		user.authenticateUser(authDetails, _this.authCallbacks(user, resolve, reject));
	});
};
...
AuthClass.prototype.createCognitoUser = function (username) {
	var userData = {
		Username: username,
		Pool: this.userPool,
	};
	userData.Storage = this._storage;
	var authenticationFlowType = this._config.authenticationFlowType;
	var user = new amazon_cognito_identity_js_1.CognitoUser(userData);
	if (authenticationFlowType) {
		user.setAuthenticationFlowType(authenticationFlowType);
	}
	return user;
};
...
```

**Comment by kirrg001** [comment]
> @TheVaporTrail currently that is not possible to do. We have an open RFC with admin auth task here.

The RFC was already closed, but it doesn't look like it covered the original request from this issue?

**Comment by magjack** [comment]
I would like to see a re enter password feature.

**Comment by jimgroome** [comment]
Hi,

We've got a need for this too. I've written a function to reset a user's MFA settings (for when they've lost their device, etc). It's only available to my app's admin users, but ideally we'd like the admin to confirm their admin password before the request is made.

**Comment by QcPerreault** [comment]
@elorzafe Any news on this 're-enter password' feature? It would be a nice +

**Comment by GabGabyz** [comment]
Hi, I need this feature too.
I already experienced Firebase and it is really more convenient than cognito
It would be really great if they can implement a reauthenticate function as firebase do as soon as possible.

**Comment by GabGabyz** [comment]
> @TheVaporTrail currently that is not possible to do. We have an open RFC with admin auth task [here](https://github.com/aws-amplify/amplify-cli/issues/766).

Hi @elorzafe,
It would be really great and more convenient if cognito can implement a reauthenticate function as firebase do.
Have you planed to realize it ?
Thanks

**Comment by awetstone56** [comment]
I'm assuming there is still no movement on this? It would be greatly needed.

**Comment by manolaz** [comment]
Please upadte this,

we now authen multiple Organization fo diddferent project!

**Comment by KvNGCzA** [comment]
This is indeed a very important feature. Would be nice if the team could implement this.

**Comment by MilosKarakas** [comment]
Hi guys, are there any chances that this might be picked up by someone at a later date? It's a critical use case for my team.

**Comment by igorrocha** [comment]
Also interested in this!

**Comment by lennartzellmer** [comment]
Very much needed feature!

**Comment by abdallahshaban557** [comment]
Hello everyone, we have informed the Cognito team of the need for this feature, however this is not on their immediate roadmap at this point in time. We will provide updates when we have progress on this issue.

**Comment by alexb148** [comment]
Also need this, very surprised to see it's not possible, and exceptionally disheartened to see this issue is 5 years old.

**Comment by Meags27** [comment]
Need this feature too, need a way to get a user to enter their password and MFA again before deleting account, changing password or email.

**Comment by codinronan** [comment]
Hate to do this but adding my own +1 here. Had an issue come up that we just can't reliably work around without first genuinely re-authenticating the user and this would have been extremely helpful to integrate into that workflow.

**Comment by abidhkm** [comment]
It is possible to re-authenticate the user using Auth.signIn API.
The implementation was as follows:
```
const reAuthenticate = async (password) => {
  await Auth.signIn(user?.getUsername(), password),
}
```
it doesn't sign-out the user session if the password is wrong.

**Comment by turakvlad** [comment]
We also require this feature. We would like to prompt our users for a password before performing sensitive actions such as changing an email, modifying MFA settings, or deleting an account.

**Comment by alexb148** [comment]
> @abidhkm It is possible to re-authenticate the user using Auth.signIn API.

This is great and seems to work. Has anybody implemented this and come across any issues, or is this a reliable working solution? Maybe this is why they haven't felt the need to implement any alternative?

**Comment by abidhkm** [comment]
> > @abidhkm It is possible to re-authenticate the user using Auth.signIn API.
> 
> This is great and seems to work. Has anybody implemented this and come across any issues, or is this a reliable working solution? Maybe this is why they haven't felt the need to implement any alternative?

using this solution for a while, have not faced any issue so far. 
I wonder why it hasn't added to the documentation or the issue closed?

**Comment by duyta7598** [comment]
> > > @abidhkm It is possible to re-authenticate the user using Auth.signIn API.
> > 
> > 
> > This is great and seems to work. Has anybody implemented this and come across any issues, or is this a reliable working solution? Maybe this is why they haven't felt the need to implement any alternative?
> 
> using this solution for a while, have not faced any issue so far. I wonder why it hasn't added to the documentation or the issue closed?

I don't think it works correctly. It will lock the account if the user enters the wrong password multiple times, which can affect the login process.
And if they use 2FA, it sends the otp code, which is a step I don't want.
We could use another function similar to this called Auth.changePassword to check if the password is wrong, but it seems like if we call it multiple times, it locks this function for a while and we cannot change the password at that time

**Comment by Meags27** [comment]
I tried calling the signIn API in V6 to re-authenticate the user and it comes back with an error saying "There is already a signed in user." Does anyone else get this?

**Comment by karolinafelus** [comment]
> I tried calling the signIn API in V6 to re-authenticate the user and it comes back with an error saying "There is already a signed in user." Does anyone else get this?

Yes, I have run into same issue today. The signIn API was changed somehow and you cannot perform another sign in while an user is authenticated.

**Comment by TeoChi-CF** [comment]
Re-authenticating is no longer possible since this change: https://github.com/aws-amplify/amplify-js/pull/12033 :(

**Comment by sudeshkumara** [comment]
> It is possible to re-authenticate the user using Auth.signIn API. The implementation was as follows:
> 
> ```
> const reAuthenticate = async (password) => {
>   await Auth.signIn(user?.getUsername(), password),
> }
> ```
> 
> it doesn't sign-out the user session if the password is wrong.


Does anyone know if we can expect support for this functionality again soon? Or can we expect a proper way to validate the password of a signed-in user?

**Comment by sudeshkumara** [comment]
> It is possible to re-authenticate the user using Auth.signIn API. The implementation was as follows:
> 
> ```
> const reAuthenticate = async (password) => {
>   await Auth.signIn(user?.getUsername(), password),
> }
> ```
> 
> it doesn't sign-out the user session if the password is wrong.


Based on the changes made here, it looks like we can't use this anymore, so does anyone know of a proper way to validate a signed-in user's password rather than using the password change API?

![Screenshot 2024-04-04 at 11 10 07](https://github.com/aws-amplify/amplify-js/assets/44965299/fee61b77-5b84-4cac-a669-5ecae7f0cdf5)

**Comment by vdastpak** [comment]
I too need to provide a "delete account" link before I can upgrade my game's Android app. The workaround I have thought of, hopefully in the meantime this request is picked up by the Cognito team, is to disable the link shortly after signing in, and to let the user know that they can send a "delete user" request only within 30 seconds or so of signing in. So a user attempting to delete their account is directed to sign out, sign back in, and then click on the link before it grays out. 
This of course is not bullet-proof but adds an extra layer of protection, at the expense of the user convenience. Any thoughts?

**Comment by johannespn** [comment]
Having the same issue

> > It is possible to re-authenticate the user using Auth.signIn API. The implementation was as follows:
> > ```
> > const reAuthenticate = async (password) => {
> >   await Auth.signIn(user?.getUsername(), password),
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > it doesn't sign-out the user session if the password is wrong.
> 
> Based on the changes made here, it looks like we can't use this anymore, so does anyone know of a proper way to validate a signed-in user's password rather than using the password change API?
> 
> ![Screenshot 2024-04-04 at 11 10 07](https://private-user-images.githubusercontent.com/44965299/319458407-fee61b77-5b84-4cac-a669-5ecae7f0cdf5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE5NDAxMTAsIm5iZiI6MTcyMTkzOTgxMCwicGF0aCI6Ii80NDk2NTI5OS8zMTk0NTg0MDctZmVlNjFiNzctNWI4NC00Y2FjLWE2NjktNWVjYWU3ZjBjZGY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI1VDIwMzY1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyNGY2ODNhMmY5ZTg2NWY5ZmVjZGQxZDQzYzM1MmZjZGI4YjFmZjI1YjM4ZDRiZWFhNDFjZTY1NDIyMGJmNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RhlhT3aMDKApNNJ41zjHmw8Pu-fx4wemNxdv7o4XyRs)

I am also wondering about this

**Comment by Mitelak** [comment]
Even though almost 7 years have passed, having this feature in the library would be great. Until this, if anyone is looking for the replacement of the `signIn` solution from v5 in the v6:

```ts

import { defaultStorage, KeyValueStorageInterface } from "aws-amplify/utils"
import { getCurrentUser, signIn } from "aws-amplify/auth"
import { cognitoUserPoolsTokenProvider } from "aws-amplify/auth/cognito"

// Anything that can store tokens temporarily
class InMemoryStorage implements KeyValueStorageInterface {
  #storage = new Map<string, string>()
  async setItem(key: string, value: string) {
    this.#storage.set(key, value)
  }
  async getItem(key: string) {
    return this.#storage.get(key) ?? null
  }
  async removeItem(key: string) {
    this.#storage.delete(key)
  }
  async clear() {
    this.#storage.clear()
  }
  getEntries() {
    return Array.from(this.#storage.entries())
  }
}

export const reauthenticate = async (password: string) => {
  const { username } = await getCurrentUser()

  const tmpStorage = new InMemoryStorage()
  cognitoUserPoolsTokenProvider.setKeyValueStorage(tmpStorage)
  
  try {
    await signIn({
      username,
      password,
      options: { authFlowType: "USER_PASSWORD_AUTH" },
    })
    tmpStorage.getEntries().forEach(([key, value]) => {
      // `defaultStorage` is the built-in localStorage strategy.     
      // If you use a different one like `CookiesStorage`, use it here instead.
      defaultStorage.setItem(key, value)
    })
  } finally {  
    // `defaultStorage` is the built-in localStorage strategy. 
    // If you use a different one like `CookiesStorage`, pass it here instead.
    cognitoUserPoolsTokenProvider.setKeyValueStorage(defaultStorage)
  }
}
```

**Comment by mathie25** [comment]
After so many years, this is still a feature we would like to see
